From ae5b6d06472f021a935f852c6ade45bac0e8554e Mon Sep 17 00:00:00 2001
Message-ID: <ae5b6d06472f021a935f852c6ade45bac0e8554e.1764242430.git.lorenzo@kernel.org>
From: Lorenzo Bianconi <lorenzo@kernel.org>
Date: Fri, 21 Nov 2025 11:40:29 +0100
Subject: [PATCH 01/14] wifi: mt76: mt7996: Fix NPU stop procedure

Fix NPU stop procedure used during device L1 SER recovery.

Fixes: 215f7c4258e1c ("wifi: mt76: mt7996: Add NPU offload support to
MT7996 driver")
Signed-off-by: Lorenzo Bianconi <lorenzo@kernel.org>
---
 mt7996/npu.c | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

diff --git a/mt7996/npu.c b/mt7996/npu.c
index 29bb735d..ac185c59 100644
--- a/mt7996/npu.c
+++ b/mt7996/npu.c
@@ -335,18 +335,19 @@ int mt7996_npu_hw_stop(struct mt7996_dev *dev)
 	for (i = 0; i < 10; i++) {
 		err = mt76_npu_get_msg(npu, 3, WLAN_FUNC_GET_WAIT_NPU_INFO,
 				       &info, GFP_KERNEL);
-		if (err)
-			continue;
+		if (!err && !info)
+			break;
 
-		if (info) {
-			err = -ETIMEDOUT;
-			continue;
-		}
+		err = -ETIMEDOUT;
+		msleep(10);
 	}
 
 	if (!err)
 		err = mt76_npu_send_msg(npu, 6,
 					WLAN_FUNC_SET_WAIT_INODE_TXRX_REG_ADDR,
 					0, GFP_KERNEL);
+	else
+		dev_err(dev->mt76.dev, "npu stop failed\n");
+
 	return err;
 }
-- 
2.52.0

